pkg/controllermanager/controller/managedseedset/add.go

```diff
@@ -0,0 +1,343 @@
+	c, err := builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&seedmanagementv1alpha1.ManagedSeedSet{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: pointer.IntDeref(r.Config.ConcurrentSyncs, 0),
+			RecoverPanic:            true,
+		}).
+		Build(r)
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &gardencorev1beta1.Shoot{}},
+		&handler.EnqueueRequestForOwner{
+			OwnerType:    &seedmanagementv1alpha1.ManagedSeedSet{},
+			IsController: true,
+		},
+		r.ShootPredicate(),
+	); err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &seedmanagementv1alpha1.ManagedSeed{}},
+		&handler.EnqueueRequestForOwner{
+			OwnerType:    &seedmanagementv1alpha1.ManagedSeedSet{},
+			IsController: true,
+		},
+		r.ManagedSeedPredicate(),
+	); err != nil {
+		return err
+	}
+
+	return c.Watch(
+		&source.Kind{Type: &gardencorev1beta1.Seed{}},
+		mapper.EnqueueRequestsFrom(mapper.MapFunc(r.MapSeedToManagedSeedSet), mapper.UpdateWithNew, c.GetLogger()),
+		r.SeedPredicate(),
+	)
+}
+
+// ShootPredicate returns the predicate for Shoot events.
+// ShootPredicate reacts only on 'CREATE','UPDATE' and 'DELETE' events. It returns true when shoot is deleted.
+// For updates,returns true if:
+// - shoot's health status changed.
+// - shoot belongs to the pending replica and it progressed from the state that caused the replica to be pending.
+func (r *Reconciler) ShootPredicate() predicate.Predicate {
+	return &shootPredicate{}
+}
+
+type shootPredicate struct {
+	ctx    context.Context
+	reader client.Reader
+}
+
+func (p *shootPredicate) InjectStopChannel(stopChan <-chan struct{}) error {
+	p.ctx = contextutil.FromStopChannel(stopChan)
+	return nil
+}
+
+func (p *shootPredicate) InjectClient(client client.Client) error {
+	p.reader = client
+	return nil
+}
+
+func (p *shootPredicate) Create(_ event.CreateEvent) bool { return true }
+
+func (p *shootPredicate) Update(e event.UpdateEvent) bool {
+	shoot, ok := e.ObjectNew.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return false
+	}
+
+	oldShoot, ok := e.ObjectOld.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return false
+	}
+
+	controllerRef := metav1.GetControllerOf(shoot)
+	if controllerRef == nil {
+		return false
+	}
+
+	managedSeedSet := &seedmanagementv1alpha1.ManagedSeedSet{}
+	if err := p.reader.Get(p.ctx, kutil.Key(shoot.Namespace, controllerRef.Name), managedSeedSet); err != nil {
+		return false
+	}
+
+	if managedSeedSet.Status.PendingReplica == nil || managedSeedSet.Status.PendingReplica.Name != shoot.Name {
+		return false
+	}
+
+	if !reflect.DeepEqual(shoot.DeletionTimestamp, oldShoot.DeletionTimestamp) || shootHealthStatus(shoot) != shootHealthStatus(oldShoot) {
+		return true
+	}
+
+	switch managedSeedSet.Status.PendingReplica.Reason {
+	case seedmanagementv1alpha1.ShootReconcilingReason:
+		return shootReconcileFailed(shoot) || shootReconcileSucceeded(shoot)
```

Comment:
I think we still need the `shoot.DeletionTimestamp != nil` check here. 
```suggestion
		return shootReconcileFailed(shoot) || shootReconcileSucceeded(shoot) || shoot.DeletionTimestamp != nil
```